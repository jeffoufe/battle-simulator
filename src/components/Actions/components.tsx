import styled from 'styled-components';
import { breakpoints } from '../../constants';

export const ActionsContainer = styled.div`
    width: 100%;
    text-align: center;
    display: flex;
    flex-direction: column;
`;

export const ActionsInfo = styled.p`
    font-size: 1em;
    text-transform: uppercase;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        font-size: 1.5em;
    }
`

export const GameOver = styled(ActionsInfo)`
    color: red;
`

export const Victory = styled(ActionsInfo)`
    color: green;
`

export const AttackButton = styled.button`
    color: white;
    background-color: red;
    padding: 10px 20px;
    border: 0px;
    margin-bottom: 20px;
    width: 50%;
    margin-left: 25%;
    font-size: 1em;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        font-size: 1.5em;
    }
`

ActionsContainer.displayName = 'ActionsContainer';
ActionsInfo.displayName = 'ActionsInfo';
GameOver.displayName = 'GameOver';
Victory.displayName = 'Victory';
AttackButton.displayName = 'AttackButton';