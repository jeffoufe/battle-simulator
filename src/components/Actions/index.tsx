import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { clickAttack, resetGame } from '../../reducers/game/actions';
import {
    ActionsContainer,
    AttackButton,
    ActionsInfo,
    GameOver,
    Victory
} from './components';

const Actions = () => {
    const dispatch = useDispatch();
    const { isTurnActive, player, monster } = useSelector((state: any) => state);
    const playerResult = player.result;
    const monsterResult = monster.result;
    const isGameFinished = player.health <= 0 || monster.health <= 0;
    const damage = player.result - monster.result;

    const onClickAttack = () => dispatch(clickAttack());
    const onResetGame = () => dispatch(resetGame());

    return (
        <ActionsContainer>
            {player.health <= 0 && (
                <GameOver>
                    Game Over
                </GameOver>
            )}
            {monster.health <= 0 && (
                <Victory>
                    You Win
                </Victory>
            )}
            {!!(playerResult && monsterResult) && (
                <ActionsInfo>
                    {!!(damage)
                        ? `You ${damage < 0 ? 'Are Hit' : 'Hit'} For ${Math.abs(damage)}`
                        : 'Draw Turn'
                    }
                </ActionsInfo>
            )}
            {!isTurnActive && (
                <AttackButton onClick={isGameFinished ? onResetGame : onClickAttack}>
                    {isGameFinished ? 'Reset': 'Attack!'}
                </AttackButton>
            )}
        </ActionsContainer>
    );
}

Actions.displayName = 'Actions';
export default Actions;