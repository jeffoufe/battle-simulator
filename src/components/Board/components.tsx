import styled from 'styled-components';
import { breakpoints } from '../../constants';

export const BoardContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  height: calc(100vh - 60px);

  @media only screen and (min-width: ${breakpoints.tablet}) {
    flex-direction: row;
    align-items: center;

    > div {
        width: calc(100% / 3);
    }
  }
`

BoardContainer.displayName = 'BoardContainer';