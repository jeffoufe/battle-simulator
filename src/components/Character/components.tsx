import styled from 'styled-components';
import { breakpoints } from '../../constants';

interface CharacterComponentProps {
    order: number
}

export const CharacterContainer = styled.div`
    display: flex;
    flex-direction: ${(props: CharacterComponentProps) => props.order ? 'column-reverse' : 'column'};

    @media only screen and (min-width: ${breakpoints.tablet}) {
        flex-direction: column-reverse;
    }
`

export const CharacterInfoDice = styled.div`
    width: 100%;
    display: flex;
    flex-direction: ${(props: CharacterComponentProps) => props.order ? 'column-reverse' : 'column'};

    @media only screen and (min-width: ${breakpoints.tablet}) {
        flex-direction: ${(props: CharacterComponentProps) => props.order ? 'row-reverse' : 'row'};;
        align-items: center;
        justify-content: space-around;
    }
`

export const CharacterInfo = styled.div`
    display: flex;
    align-items: center;
    justify-content: space-evenly;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        flex-direction: ${(props: CharacterComponentProps) => props.order ? 'row-reverse' : 'row'};
        width: calc(100% - 70px);
    }
`

export const CharacterImage = styled.img`
    width: 85px;
    height: 85px;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        width: 150px;
        height: 150px;
    }

    @media only screen and (min-width: ${breakpoints.laptop}) {
        width: 200px;
        height: 200px;
    }
`;

export const CharacterName = styled.div`
    text-align: center;
    ${(props: CharacterComponentProps) => props.order ? 'margin-top: 10px;' : 'margin-bottom: 10px;'}

    @media only screen and (min-width: ${breakpoints.tablet}) {
        margin-top: 30px;
        font-size: 2em;
    }
`

export const CharacterDice = styled.div`
    ${(props: CharacterComponentProps) => props.order ? 'margin-bottom' : 'margin-top'}: 10px;
    > .dice {
        text-align: center;
    }

    @media only screen and (min-width: ${breakpoints.tablet}) {
        > .dice {
            display: flex;
            flex-direction: column;

            .die-container {
                display: block !important;
                margin: 20px !important;
            }
        }
    }
`