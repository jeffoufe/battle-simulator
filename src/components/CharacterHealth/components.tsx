import styled from 'styled-components';
import { getHealthColor } from './utils';
import { breakpoints } from '../../constants';

interface CharacterHealthIndicatorProps {
    health: number
}

export const CharacterHealthContainer = styled.div`
    display: flex;
    flex-direction: column;
`

export const CharacterHealthIndicatorContainer = styled.div`
    width: 40px;
    height: ${60}px;
    border: 1px solid black;
    position: relative;

    @media only screen and (min-width: ${breakpoints.tablet}) {
        height: 150px;
    }

    @media only screen and (min-width: ${breakpoints.laptop}) {
        height: 200px;
    }
`

export const CharacterHealthIndicator = styled.div`
    width: 40px;
    height: ${(props: CharacterHealthIndicatorProps) => props.health}%;
    background-color: ${(props: CharacterHealthIndicatorProps) => getHealthColor(props.health)};
    position: absolute;
    bottom: 0;
`

export const CharacterHealthPoints = styled.span`
    margin-top: 5px;
    margin-bottom: 0;
    text-align: center;
`