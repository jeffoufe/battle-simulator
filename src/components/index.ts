export { default as Title } from './Title';
export { default as Character } from './Character';
export { default as Actions } from './Actions';
export { default as Board } from './Board';