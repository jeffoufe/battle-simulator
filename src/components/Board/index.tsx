import React from 'react';
import { BoardContainer } from './components';
import { Title, Character, Actions } from '..';
import WARRIOR_IMAGE from '../../assets/warrior.png';
import MONSTER_IMAGE from '../../assets/monster.png';

const Board = () => (
  <>
    <Title>Battle Simulator</Title>
    <BoardContainer>
      <Character
        image={WARRIOR_IMAGE}
        order={0}
      />
      <Actions />
      <Character
        image={MONSTER_IMAGE}
        order={1}
      />
    </BoardContainer>
  </>
)

export default Board;
