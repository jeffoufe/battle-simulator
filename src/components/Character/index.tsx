import React, { useRef, useEffect, MutableRefObject } from 'react';
import CharacterHealth from '../CharacterHealth';
import ReactDice from 'react-dice-complete';
import 'react-dice-complete/dist/react-dice-complete.css';
import { useDispatch, useSelector } from 'react-redux';
import { MONSTER_ROLL, END_TURN } from '../../reducers/game/constants';

import { 
    CharacterImage, 
    CharacterContainer, 
    CharacterInfoDice,
    CharacterInfo,
    CharacterDice,
    CharacterName
} from './components';

interface CharacterProps {
    image: string,
    order: number,
}

interface DiceRef extends MutableRefObject<null> {
    rollAll: () => void
}


const Character = ({ image, order }: CharacterProps) => {
    const dispatch = useDispatch();
    const { health, isRolling } = useSelector((state: any) => state[order ? 'monster' : 'player'])
    let diceRef = useRef<DiceRef>(null);

    useEffect(() => {
        if (isRolling) {
            diceRef.rollAll();
        }
    }, [isRolling])

    const onRollDone = (result: number) => {
        if (!order) {
            dispatch({
                type: MONSTER_ROLL,
                payload: { result }
            })
        } else {
            dispatch({
                type: END_TURN,
                payload: { result }
            })
        }
    }

    return (
        <CharacterContainer order={order}>
            <CharacterName order={order}>
                {order ? 'Monster' : 'Player'}
            </CharacterName>
            <CharacterInfoDice order={order}>
                <CharacterInfo order={order}>
                    <CharacterImage src={image} />
                    <CharacterHealth health={health} />
                </CharacterInfo>
                <CharacterDice order={order}>
                    <ReactDice
                        numDice={2}
                        faceColor='#FFFFFF'
                        dotColor='#000000'
                        rollTime={1}
                        dieSize={30}
                        outline
                        ref={(dice: DiceRef) => diceRef = dice}
                        rollDone={onRollDone}
                    />
                </CharacterDice> 
            </CharacterInfoDice>
        </CharacterContainer>
    )
}

Character.displayName = 'Character';
export default Character;