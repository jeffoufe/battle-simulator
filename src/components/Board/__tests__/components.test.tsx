import React from 'react';
import { shallow } from 'enzyme';
import { BoardContainer } from '../components'

describe('Board Components', () => {
    it('BoardContainer', () => {
        const wrapper = shallow(
            <BoardContainer />
        );
        expect(wrapper).toMatchSnapshot();
    })
})