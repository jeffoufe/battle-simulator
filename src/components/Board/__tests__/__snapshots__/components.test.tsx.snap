// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Components BoardContainer 1`] = `
<div
  className="sc-AxjAm iyergB"
/>
`;
